src/Psalm/Internal/Provider/ReturnTypeProvider/DirnameReturnTypeProvider.php

```diff
@@ -0,0 +1,64 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Psalm\Internal\Provider\ReturnTypeProvider;
+
+use PhpParser;
+use Psalm\Internal\Analyzer\Statements\Expression\IncludeAnalyzer;
+use Psalm\Plugin\EventHandler\Event\FunctionReturnTypeProviderEvent;
+use Psalm\Plugin\EventHandler\FunctionReturnTypeProviderInterface;
+use Psalm\Type\Atomic\TLiteralString;
+use Psalm\Type\Union;
+
+use function count;
+use function dirname;
+
+class DirnameReturnTypeProvider implements FunctionReturnTypeProviderInterface
+{
+    /**
+     * @return array<lowercase-string>
+     */
+    public static function getFunctionIds(): array
+    {
+        return ['dirname'];
+    }
+
+    public static function getFunctionReturnType(FunctionReturnTypeProviderEvent $event): ?Union
+    {
+        $call_args = $event->getCallArgs();
+        if (count($call_args) === 0) {
+            return null;
+        }
+
+        $dir_level = 1;
+
+        if (isset($call_args[1])) {
+            if ($call_args[1]->value instanceof PhpParser\Node\Scalar\LNumber) {
```

Comment:
Please extract this literal value from the node type provider, not directly from the statement

Reply:
True :) This will allow Psalm to infer the number even if it comes from a variable, a function call or any other expression